design-proposals/klip-0-add-protobuf-support.md

```diff
@@ -0,0 +1,77 @@
+# KLIP 0 - Add Protobuf SerDe Support
+
+Author: [Stephen Powis](http://github.com/crim)
+
+Release target: 5.1
+
+Status: In Discussion 
+
+Discussion: _link to the design discussion PR_
+
+## tl;dr
+
+Add support to KSQL to process messages stored using the [ProtocolBuffers](https://developers.google.com/protocol-buffers/) serialization format.   
+
+## Motivation and background
+
+[ProtocolBuffers](https://developers.google.com/protocol-buffers/) is a language and platform neutral serialization format that is commonly used with Apache Kafka. Adding out-of-the-box support for this message format would expand the environments in which KSQL can be utilized.
+
+## What is in scope
+
+Adding SerDe support for ProtocolBuffers Version 3 to KSQL.
+
+## What is not in scope
+
+Supporting streams/topics containing multiple ProtocolBuffer schema types.  _See [Design](#Design) section below for more details._
+
+## Value/Return
+
+Out of the box support for running KSQL against Apache Kafka topics containing messages encoded with ProtocolBuffers.
+
+## Public APIS
+
+Proposed extending the `CREATE STREAM` grammar to include a new _value_format_ type **PROTOBUF** as well as adding a new _tableProperty_ named **protobuf_class**.
+
+Example:
+```genericsql
+ksql> CREATE STREAM stream_name (....) WITH (kafka_topic='TopicName', value_format='PROTOBUF', protobuf_class='org.whatever.protobufs.MyProtoBuf$EncodedMessageType');
+```
+
+## Design  
+
+Before digging into the design, it's important to understand how ProtocolBuffers work. A ProtocolBuffer encoded message does **not** include the schema definition. A consumer attempting to deserialize a ProtocolBuffer encoded message must know in advance the schema of that message. Because ProtocolBuffer messages lack this identifying information, services like SchemaRegistry do not exist for ProtocolBuffers.   
```

Comment:
It's worth noting that an Avro encoded message also does **not** include the schema definition.  The Schema Registry works by associating a schema with a topic name.  This is currently an Avro schema, but could equally be a PB schema.

The main limitation with PB is that, unlike Avro, it has no runtime capability to take a PB schema and decode an encoded message. It requires a compile-time generated Java class.

Reply:
doing this at runtime is certainly possible! tooling/library support is weak currently, but i think it is going to improve. 

in go there is an excellent package: https://github.com/jhump/protoreflect

but also for java, some work has been done: https://github.com/os72/protobuf-dynamic

Reply:
> The main limitation with PB is that, unlike Avro, it has no runtime capability to take a PB schema and decode an encoded message. It requires a compile-time generated Java class.

To me that is rather a feature since I rather have stable execution times than some obscure magic generating code.